engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java

```diff
@@ -0,0 +1,353 @@
+import com.oracle.truffle.api.interop.*;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import java.time.*;
+import org.enso.interpreter.arrow.LogicalLayout;
+
+@ExportLibrary(InteropLibrary.class)
+public final class ArrowFixedSizeArrayBuilder implements TruffleObject {
+  private final ByteBufferDirect buffer;
+  private final LogicalLayout unit;
+  private final int size;
+  private int index;
+
+  private boolean sealed;
+
+  public ArrowFixedSizeArrayBuilder(int size, LogicalLayout unit) {
+    this.size = size;
+    this.unit = unit;
+    this.buffer = ByteBufferDirect.forSize(size, unit);
+    this.index = 0;
+    this.sealed = false;
+  }
+
+  public LogicalLayout getUnit() {
+    return unit;
+  }
+
+  @ExportMessage
+  public boolean hasArrayElements() {
+    return true;
+  }
+
+  @ExportMessage
+  public boolean hasMembers() {
+    return true;
+  }
+
+  @ExportMessage
+  public boolean isMemberInvocable(String member) {
+    return switch (member) {
+      case "build", "append" -> !this.sealed;
+      default -> false;
+    };
+  }
+
+  @ExportMessage
+  Object getMembers(boolean includeInternal) throws UnsupportedMessageException {
+    return this;
+  }
+
+  @ExportMessage
+  @ImportStatic(LogicalLayout.class)
+  static class InvokeMember {
+    @Specialization(guards = "isDateLayout(receiver.getUnit())")
+    public static Object doInvokeDate(
+        ArrowFixedSizeArrayBuilder receiver,
+        String name,
+        Object[] args,
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException, UnknownIdentifierException {
+      switch (name) {
+        case "build":
+          receiver.sealed = true;
+          return new ArrowFixedArrayDate(receiver.buffer, receiver.unit);
+        case "append":
+          if (receiver.sealed) {
+            throw UnsupportedMessageException.create();
+          }
+          var current = receiver.index;
+          try {
+            iop.writeArrayElement(receiver, current, args[0]);
+          } catch (UnsupportedTypeException | InvalidArrayIndexException e) {
+            throw UnsupportedMessageException.create(e);
+          }
+          receiver.index += 1;
+          return NullValue.get();
+        default:
+          throw UnknownIdentifierException.create(name);
+      }
+    }
+
+    @Specialization(guards = "!isDateLayout(receiver.getUnit())")
+    public static Object doInvokeInt(
+        ArrowFixedSizeArrayBuilder receiver,
+        String name,
+        Object[] args,
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException, UnknownIdentifierException {
+      switch (name) {
+        case "build":
+          receiver.sealed = true;
+          return new ArrowFixedArrayInt(receiver.buffer, receiver.unit);
```

Comment:
Is `doInvokeDate` and `doInvokeInt` the same except this line? Then it'd be better to avoid the duplicated code for the sake of future maintenance.